### ISSUE TITLE
Cross-Origin Request Blocked

### ISSUE THREAD

**Issue by wontone18** [fruitcake/laravel-cors#489]
Hi

I tried all the way but its blocking. I am using a laravel 7.2 version. And my front end is react. I tested this with fetch and axio but it is not working.

Cross-Origin Request Blocked: The Same Origin Policy disallows reading the remote resource at http://localhost/smarttalkonlineedu/webapps/api/public/api/loginin. (Reason: CORS header ‘Access-Control-Allow-Origin’ missing).

**Comment by wontone18** [comment]
My error now is CORS multiple origin not allowd.

**Comment by wontone18** [comment]
thanks its fix for my side. if any body facing the issue then use header pattern. This resolve my issue

<?php

return [

    /*
    |--------------------------------------------------------------------------
    | Cross-Origin Resource Sharing (CORS) Configuration
    |--------------------------------------------------------------------------
    |
    | Here you may configure your settings for cross-origin resource sharing
    | or "CORS". This determines what cross-origin operations may execute
    | in web browsers. You are free to adjust these settings as needed.
    |
    | To learn more: https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS
    |
    */

    'paths' => ['api/*'],

    'allowed_methods' => ['GET, POST, PUT, PATCH, DELETE, OPTIONS'],

    'allowed_origins' => ['*'],

    'allowed_origins_patterns' => [],

    'allowed_headers' => ['Authorization,Accept,Origin,DNT,X-CustomHeader,Keep-Alive,User-Agent,X-Requested-With,If-Modified-Since,Cache-Control,Content-Type,Content-Range,Range'],

    'exposed_headers' => [],

    'max_age' => 0,

    'supports_credentials' => false,

];

**Comment by barryvdh** [comment]
Did wildcard for allowed_headers not work?

**Comment by wontone18** [comment]
no barry it was not working. After adding those line it is now working from your package.

**Comment by barryvdh** [comment]
Which browser were you using?

**Comment by wontone18** [comment]
> Which browser were you using?

i had tested this in firefox and chrome both.

**Comment by wontone18** [comment]
one more thing i will also share here for people. Request and Response should always be in json.For instance if its like somebody sending a json value to back end but in return response is without json then browser will also block the url. 

we must ensure about this request and response should be in json.

**Comment by thalesbarbosab** [comment]
Hello guys.

I was have a same issue when i deploy to production. 
Just solve for me when i run that comand:

`php artisan config:clear`

Good luck.

**Comment by iBotPeaches** [comment]
I'm confused how I've been breaking this for years, but it appears the spec has caught with browsers in the enforcement. Probably happened around when Chrome started enforcing cookie lax/none.

It appears if you run credentials (whether cookie) or a Authorization header. You can no longer use the wildcard for `allowed_headers`.

I found this in the spec.

> For `Access-Control-Expose-Headers`, `Access-Control-Allow-Methods`, and `Access-Control-Allow-Headers` response headers the value `*` counts as a wildcard for requests **without credentials**. 

https://fetch.spec.whatwg.org/#http-access-control-expose-headers

---

Or if you prefer MDN - https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Expose-Headers

> The value "*" only counts as a special wildcard value for requests without credentials (requests without HTTP cookies or HTTP authentication information). In requests with credentials, it is treated as the literal header name "*" without special semantics.

So my config dropped all wildcards.

---

```
    'supports_credentials' => true,
    'allowed_headers' => [
        'Accept',
        'Authorization',
        'Content-Type',
        'Cookie',
        'DNT',
        'Origin',
        'User-Agent',
        'X-Requested-With',
    ],
    'exposed_headers' => [
        'Accept',
        'Authorization',
        'Origin',
        'Content-Type',
        'X-Requested-With',
    ],
```

---

Now everything I've tested is now working again. Hope that helps someone. I might try and push some doc updates when I grab a moment. Hope that helps someone for now.

**Comment by barryvdh** [comment]
Hmm but we should transform the headers to just return the requested value: https://github.com/asm89/stack-cors/blob/88aaf08c68a66647309b7535f55591dc785b29ed/src/CorsService.php#L178-L187
Will see if I can reproduce

**Comment by iBotPeaches** [comment]
> Hmm but we should transform the headers to just return the requested value: https://github.com/asm89/stack-cors/blob/88aaf08c68a66647309b7535f55591dc785b29ed/src/CorsService.php#L178-L187
> Will see if I can reproduce

I see, so the array with asterik gets to stack-cors package, which is normalized to true. Stack-cors then detects that true value and pulls the header value from the request - `Access-Control-Request-Headers` back into the response. If you don't use asterik/true, it just implodes into comma delimited.

So my preflight on app is
```
Access-Control-Request-Headers: content-type
```

stack-cors detects my config as asterik, changes to true, copies that and returns it. Which is now invalid because it dropped `Origin` 

Hardcode the array since credentials and I get back in response
```
Access-Control-Allow-Headers: accept, authorization, content-type, cookie, dnt, origin, user-agent, x-requested-with
```

No issue.

**Comment by iBotPeaches** [comment]
> Are you using 2.0.1 of stack-cors?

            "name": "asm89/stack-cors",
            "version": "v2.0.1",

            "name": "fruitcake/laravel-cors",
            "version": "v2.0.1",

**Comment by barryvdh** [comment]
But it doesn't drop Authorization, that header is not requested by your `Access-Control-Request-Headers`?

**Comment by iBotPeaches** [comment]
> But it doesn't drop Authorization, that header is not requested by your `Access-Control-Request-Headers`?

Correct, I'm not a true master at the understanding yet of all this CORS stuff. I wasn't aware till now that the asterisk is just converted behind scenes to whatever the request passed.

I am a bit confused though, if stack-cors just pulls `Access-Control-Request-Headers` if an asterisk is used. Am I expected to use that header in client side side code for all my allowed headers knowing that it needs to be expanded server side and returned?

Some of my legacy bad code is as ugly as

```
    $.ajax({
        type: 'POST',
        url: url
        data: JSON.stringify({}),
        headers: {
            'Accept': 'application/json',
            'Content-Type': 'application/json'
        },
        xhrFields: {withCredentials: true},
        success: function (response) {
          //
        },
    });
```

So I never set any headers outside of those.

**Comment by barryvdh** [comment]
I would expect that the browser would take care of the Request-Headers values, you wouldn't have to do that manually.

**Comment by iBotPeaches** [comment]
> I would expect that the browser would take care of the Request-Headers values, you wouldn't have to do that manually.

I'll do some low level nginx logging to get everything to see if I can't get a better understanding around this and loop back. For now, my manual setting has resolved the issues that started cropping up for isolated users/environments.

**Comment by lvxianchao** [comment]
> 'Authorization,Accept,Origin,DNT,X-CustomHeader,Keep-Alive,User-Agent,X-Requested-With,If-Modified-Since,Cache-Control,Content-Type,Content-Range,Range'

it dosen't work for me.

**Comment by KieronWiltshire** [comment]
I'm pretty sure this is down to the failing check of the asterisk wildcard.

**Comment by ALuciferDK** [comment]
cors 2.0 The second request is intercepted when the Axios request is passed,It could be cross domain,And now it's impossible to make a cross domain request anyway,It has been modified in the above way, but it still doesn't work,this is my Chrome error
![Uploading error.png…]()

**Comment by ALuciferDK** [comment]
I found the reason, because I added a custom header remember token. I want to ask how to add this into the allowed scope of the header

**Comment by ALuciferDK** [comment]
Yes, because the configuration is too confusing, and then it's OK to just use CORS

**Comment by trainoasis** [comment]
For me, none of the above helps, trying for a whole day now. Simple API GET request via Axios (VueJS). Am on localhost, and my API is on localhost, but with a domain set such as myapi.my. Any tips?

**Comment by maxacarvalho** [comment]
Hello @barryvdh 

I'm sorry if the issue I'm about to report does not relate to this one but I'm just out of options, went all over the internet and I can't crack this.

I'm using Laravel Sanctum and Vue CLI. Everything works fine on Chrome and Safari, both normal and incognito windows.  
I already refreshed my cache and configurations many times.  
This a local development, not in production yet.

Here's the Laravel `config/cors.php` file contents:

```php
<?php

return [

    /*
    |--------------------------------------------------------------------------
    | Cross-Origin Resource Sharing (CORS) Configuration
    |--------------------------------------------------------------------------
    |
    | Here you may configure your settings for cross-origin resource sharing
    | or "CORS". This determines what cross-origin operations may execute
    | in web browsers. You are free to adjust these settings as needed.
    |
    | To learn more: https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS
    |
    */

    'paths' => ['api/*', '/login', '/logout', '/sanctum/csrf-cookie'],

    'allowed_methods' => ['*'],

    'allowed_origins' => [
        'https://web.nicksaude.test',
        'http://web.nicksaude.test',
        'https://web.nicksaude.test:8000',
        'http://web.nicksaude.test:8000',
        'https://api.nicksaude.test',
        'http://api.nicksaude.test',
    ],

    'allowed_origins_patterns' => [],

    'allowed_headers' => [
        'Accept',
        'Authorization',
        'Content-Type',
        'Cookie',
        'cookie',
        'DNT',
        'origin',
        'User-Agent',
        'X-Requested-With',
        'x-xsrf-token',
        'X-XSRF-TOKEN',
        'referer',
        'Referer',
    ],

    'exposed_headers' => [
        'Accept',
        'Authorization',
        'Origin',
        'Content-Type',
        'X-Requested-With',
    ],

    'max_age' => 0,

    'supports_credentials' => true,
];
```

Here's my Axios config in a nutshell
```javascript
...
const api = axios.create({
  baseURL: process.env.VUE_APP_API_URL,
  withCredentials: true,
  headers: {
    Accept: 'application/json',
    'Content-Type': 'application/json',
  },
});
...
```

When I use Firefox (I tried developer and standard editions and I reset the configurations) here's what I get back:

Console log:
```
Cross-Origin Request Blocked: The Same Origin Policy disallows reading the remote resource at https://api.nicksaude.test/api/admin/users/me. (Reason: CORS request did not succeed).
```

Raw Request Headers:
```
GET /api/admin/users/me undefined
Host: api.nicksaude.test
User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10.16; rv:83.0) Gecko/20100101 Firefox/83.0
Accept: application/json
Accept-Language: en-US,en;q=0.5
Accept-Encoding: gzip, deflate, br
Origin: https://web.nicksaude.test:8000
Connection: keep-alive
Referer: https://web.nicksaude.test:8000/login
```

**Comment by oluudeh** [comment]
I am having the same issue on Laravel 8.

**Comment by wixys** [comment]
In my case, I'm using a multi-tenancy system, this package required some special settings in the RouteServiceProvider and actually I'm using a routes/tenant.php file to manage routes api, so I've changed in the config/cors.php this value:
from: 
'paths' => ['api/*'] 
to  
'paths' => ['/*']

Working fine!

**Comment by vangelis183** [comment]
As long a wildcard is being used in paths, it doesn't work and returns CORS Errors.

**Comment by gniuslab** [comment]
Do not use this format in your routes https://www.domain.com/api/login/
The last slash returns CORS, remove it and it will work.

**Comment by blu-towers** [comment]
Same issue as @gniuslab 

The default '/' route with a header of Content-Type: application/json caused CORS errors but only when I sent Content-Type.

If i change the route to '/search' it works.

**Comment by sergiogithub** [comment]
Yes, same as @gniuslab 

The last '/' is generating CORS error.

Thanks man.

**Comment by mhnaufal** [comment]
> In my case, I'm using a multi-tenancy system, this package required some special settings in the RouteServiceProvider and actually I'm using a routes/tenant.php file to manage routes api, so I've changed in the config/cors.php this value:
> from:
> 'paths' => ['api/_'] to 'paths' => ['/_']
> 
> Working fine!

yes, it's working on me

**Comment by louisfitz** [comment]
This point is valid: return Responses should configured as a JSON response, rather than just returning the value, other wise it will also be blocked by CORS policy.

in Laravel, you sholud return responses like this: 

-- return response()->Json($value); --, rather than just -- return $value --

**Comment by linehammer** [comment]
The Same Origin Policy (SOP) is a security measure standardized among browsers. It is needed to prevent Cross-Site Request Forgery (CSRF). The "Origin" mostly refers to a "Domain". Same Origin Policy prevents different origins (domains) from interacting with each other, to prevent attacks such as CSRF (Cross Site Request Forgery) through such requests, like AJAX. In other words, the browser would not allow any site to make a request to any other site. Without Same Origin Policy , any web page would be able to access the DOM of other pages.

This SOP (Same Origin Policy) exists because it is too easy to inject a link to a javascript file that is on a different domain. This is actually a security risk ; you really only want code that comes from the site you are on to execute and not just any code that is out there.

If you want to bypass that restriction when fetching the contents with fetch API or XMLHttpRequest in [javascript](http://net-informations.com/js/iq/default.htm), you can use a proxy server so that it sets the header Access-Control-Allow-Origin to *.

If you need to enable CORS on the server in case of localhost, you need to have the following on request header.

`Access-Control-Allow-Origin: http://localhost:9999`

**Comment by SWS-5007** [comment]
I am Deploying the Laravel Project on Vercel.
And then, I am getting this CORS Error.

> Response to preflight request doesn't pass access control check: It does not have HTTP ok status.

Here is my Vercel.json file on root of Laravel Project.

```
<?php
namespace App\Http\Middleware;

use Closure;
use Illuminate\Http\Request;

class Cors
{
    public function handle(Request $request, Closure $next)
    {
        header("Access-Control-Allow-Origin: https://businesscard-frontend.netlify.app/");
        // ALLOW OPTIONS METHOD
        $headers = [
            'Access-Control-Allow-Methods'=> 'POST, GET, OPTIONS, PUT, DELETE',
            'Access-Control-Allow-Headers'=> 'Origin, Content-Type, X-Auth-Token, Authorization, X-Requested-With, x-xsrf-token'
        ];
        if($request->getMethod() == "OPTIONS") {
            // The client-side application can set only headers allowed in Access-Control-Allow-Headers
            return Response::make('OK', 200, $headers);
        }
        $response = $next($request);
        foreach($headers as $key => $value)
            $response->header($key, $value);
        return $response;
    }
}
```

What is Wrong in my code?
Please help me!
